{"cells":[{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":304},"executionInfo":{"elapsed":3039,"status":"error","timestamp":1723144946331,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"St1Xp_IiILCr","outputId":"efb4d3c6-ea0a-4d5b-cdae-679d7869c110"},"outputs":[{"output_type":"error","ename":"MessageError","evalue":"Error: credential propagation was unsuccessful","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mMessageError\u001b[0m                              Traceback (most recent call last)","\u001b[0;32m<ipython-input-1-d5df0069828e>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mgoogle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolab\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mdrive\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdrive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/content/drive'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/drive.py\u001b[0m in \u001b[0;36mmount\u001b[0;34m(mountpoint, force_remount, timeout_ms, readonly)\u001b[0m\n\u001b[1;32m     98\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmountpoint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mforce_remount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m120000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreadonly\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m   \u001b[0;34m\"\"\"Mount your Google Drive at the specified mountpoint path.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 100\u001b[0;31m   return _mount(\n\u001b[0m\u001b[1;32m    101\u001b[0m       \u001b[0mmountpoint\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m       \u001b[0mforce_remount\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mforce_remount\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/drive.py\u001b[0m in \u001b[0;36m_mount\u001b[0;34m(mountpoint, force_remount, timeout_ms, ephemeral, readonly)\u001b[0m\n\u001b[1;32m    131\u001b[0m   )\n\u001b[1;32m    132\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mephemeral\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 133\u001b[0;31m     _message.blocking_request(\n\u001b[0m\u001b[1;32m    134\u001b[0m         \u001b[0;34m'request_auth'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m         \u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m'authType'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'dfs_ephemeral'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_message.py\u001b[0m in \u001b[0;36mblocking_request\u001b[0;34m(request_type, request, timeout_sec, parent)\u001b[0m\n\u001b[1;32m    174\u001b[0m       \u001b[0mrequest_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparent\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparent\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpect_reply\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    175\u001b[0m   )\n\u001b[0;32m--> 176\u001b[0;31m   \u001b[0;32mreturn\u001b[0m \u001b[0mread_reply_from_input\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_sec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_message.py\u001b[0m in \u001b[0;36mread_reply_from_input\u001b[0;34m(message_id, timeout_sec)\u001b[0m\n\u001b[1;32m    101\u001b[0m     ):\n\u001b[1;32m    102\u001b[0m       \u001b[0;32mif\u001b[0m \u001b[0;34m'error'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mreply\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 103\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mMessageError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreply\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'error'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    104\u001b[0m       \u001b[0;32mreturn\u001b[0m \u001b[0mreply\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    105\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mMessageError\u001b[0m: Error: credential propagation was unsuccessful"]}],"source":["from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":2,"status":"aborted","timestamp":1723144946332,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"84-3M1ijeB_1"},"outputs":[],"source":["# @title Data Reading\n","import pandas as pd\n","import os\n","from pandas import ExcelFile\n","import matplotlib.pyplot as plt\n","import seaborn as sns\n","from random import sample\n","import numpy as np\n","from statsmodels.tsa.stattools import adfuller\n","from statsmodels.stats.multitest import multipletests\n","import warnings\n","from scipy.spatial.distance import pdist, squareform\n","from scipy.stats import entropy\n","from scipy.signal import savgol_filter\n","from scipy.spatial.distance import euclidean\n","from sklearn.preprocessing import StandardScaler\n","\n","from sklearn.decomposition import PCA\n","from sklearn.metrics import pairwise_distances\n","from mpl_toolkits.mplot3d import Axes3D\n","import matplotlib.patches as mpatches\n","from scipy.stats import gmean\n","import math\n","\n","gene_count_path ='/content/drive/MyDrive/MITResearch/IBD_13/gene_count_preprocessed.csv'\n","metaphlan_path = '/content/drive/MyDrive/MITResearch/IBD_13/metaphlan_preprocessed.csv'\n","\n","gene_count_df = pd.read_csv(gene_count_path)\n","metaphlan_df = pd.read_csv(metaphlan_path)\n","\n","print(gene_count_df.shape)\n","print(metaphlan_df.shape)"]},{"cell_type":"code","source":["\n","extended_colors = [\n","    '#66c2a5',  # Teal\n","    '#ff7f00',  # Orange\n","    '#8da0cb',  # Light Blue\n","    '#e78ac3',  # Pink\n","    '#a6d854',  # Light Green\n","    '#e5c494',  # Tan\n","    '#b3b3b3',  # Gray\n","    '#80b1d3',  # Sky Blue\n","    '#bebada',  # Lavender\n","    '#fb8072',  # Light Red\n","    '#7fc97f',  # Medium Green\n","    '#d9d9d9',  # Light Gray\n","    '#ffed6f',  # Pale Yellow\n","    '#b15928',  # Brown\n","    '#fccde5',  # Pale Pink\n","    '#8dd3c7',  # Pale Teal\n","    '#ffffb3',  # Very Pale Yellow\n","    '#bc80bd'   # Purple\n","]\n","\n","# Set default color cycle\n","plt.rcParams['axes.prop_cycle'] = plt.cycler(color=extended_colors)\n","\n","# Set other default styles\n","#plt.rcParams['figure.facecolor'] = '#f0f0f0'\n","plt.rcParams['axes.facecolor'] = 'white'\n","plt.rcParams['grid.color'] = '#cccccc'\n","plt.rcParams['grid.linestyle'] = '--'\n","plt.rcParams['grid.alpha'] = 0.5\n","\n","# Adjust grid frequency\n","plt.rcParams['axes.grid'] = False\n","plt.rcParams['axes.grid.which'] = 'major'\n","plt.rcParams['axes.grid.axis'] = 'both'\n","\n","\n","# Set default line width\n","plt.rcParams['lines.linewidth'] = 1"],"metadata":{"id":"26mdoqpzQlYR","executionInfo":{"status":"aborted","timestamp":1723144946332,"user_tz":240,"elapsed":2,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":2,"status":"aborted","timestamp":1723144946332,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"zgkBnIaFkPpD"},"outputs":[],"source":["# @title count columns with non zero\n","def count_columns_with_nonzero(df, excluded_columns):\n","    # Exclude the specified columns\n","    df_filtered = df.drop(columns=excluded_columns)\n","\n","    # Count non-zero values in each column\n","    non_zero_counts = df_filtered.astype(bool).sum(axis=0)\n","\n","    # Get the number of samples\n","    num_samples = len(df_filtered)\n","\n","    # Find columns where the count of non-zero values is greater than 10% of the number of samples\n","    more_than_10_percent = non_zero_counts > (0.2 * num_samples)\n","\n","    # Count these columns\n","    count = more_than_10_percent.sum()\n","\n","    return count\n","\n","excluded_columns = ['patient_id', 'week', 'Flare_status']\n","result = count_columns_with_nonzero(gene_count_df, excluded_columns)\n","print(f'Number of columns with more than 20% non-zero values in gene_count_df: {result}')\n","\n","result = count_columns_with_nonzero(metaphlan_df, excluded_columns)\n","print(f'Number of columns with more than 20% non-zero values in metaphlan_df: {result}')"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":2,"status":"aborted","timestamp":1723144946332,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"P7r4Kv6iux1X"},"outputs":[],"source":["# @title flag_first_flare_weeks function\n","\n","def flag_first_flare_weeks(df):\n","    # Identify the rows where flare starts\n","    df['is_flare'] = (df['Flare_status'] == 'During_flare') | (df['Flare_status'] == 'During_flare_2')\n","\n","    # Sort by patient and week to ensure the chronological order\n","    df.sort_values(by=['patient_id', 'week'], inplace=True)\n","\n","    # Mark each flare start for each patient\n","    df['Flare_start'] = (df['is_flare']) & (df['is_flare'] != df['is_flare'].shift(1))\n","\n","    # Convert boolean to integer (1 for True, 0 for False)\n","    df['Flare_start'] = df['Flare_start'].astype(int)\n","\n","    flare_start = df.pop('Flare_start')\n","    # Now, insert it at the desired position\n","    df.insert(3, 'Flare_start', flare_start)\n","\n","    # Drop helper columns if they are no longer needed\n","    df.drop('is_flare', axis=1, inplace=True)\n","\n","    return df"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":3674,"status":"aborted","timestamp":1723144946332,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"z0Y4Z7kYu6wJ"},"outputs":[],"source":["# @title align_flare_start_function\n","def align_flare_start_week(df, flare_column='Flare_start', week_column='week', patient_column='patient_id', target_week=0):\n","    \"\"\"\n","    Adjusts the week numbers for each patient in the dataframe such that the first occurrence\n","    of Flare_start=1 (that is not the first sample for the patient) occurs in the same target week for all patients who have a flare start.\n","\n","    Args:\n","    df (pd.DataFrame): DataFrame containing the data.\n","    flare_column (str): Column name indicating the flare start.\n","    week_column (str): Column name indicating the week number.\n","    patient_column (str): Column name indicating the patient ID.\n","    target_week (int): The week number to which the first flare start should be aligned.\n","\n","    Returns:\n","    pd.DataFrame: Modified DataFrame with adjusted week numbers.\n","    \"\"\"\n","    df = df.copy()\n","    adjusted_dfs = []\n","\n","    # Add a column indicating if the patient had a flare or not\n","    df['had_flare'] = df.groupby(patient_column)[flare_column].transform('max')\n","\n","    for patient in df[patient_column].unique():\n","        patient_data = df[df[patient_column] == patient].copy()\n","        first_week = patient_data[week_column].min()\n","\n","        if patient_data['had_flare'].iloc[0] == 1:\n","            # Identify flare weeks and remove the first week if it's a flare\n","            flare_weeks = patient_data[patient_data[flare_column] == 1][week_column]\n","            flare_weeks = flare_weeks[flare_weeks > first_week]  # ignore flare at the first recorded week\n","\n","            if not flare_weeks.empty:\n","                flare_week = flare_weeks.min()  # first flare that's not at the first week\n","                shift = target_week - flare_week\n","                patient_data[week_column] += shift\n","\n","        adjusted_dfs.append(patient_data)\n","\n","    aligned_df = pd.concat(adjusted_dfs, ignore_index=True)\n","\n","    # Normalize to ensure there are no negative weeks\n","    min_week = aligned_df[week_column].min()\n","    if min_week < 0:\n","       aligned_df.loc[aligned_df['had_flare'] == 1, week_column] -= min_week\n","\n","    # Remove the temporary 'had_flare' column before returning\n","    aligned_df.drop('had_flare', axis=1, inplace=True)\n","\n","    return aligned_df"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":1,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"9WVJHAt1vx6F"},"outputs":[],"source":["# @title analyze representation and variation\n","\n","def analyze_representation_and_variation(df, start_col, category_name):\n","    \"\"\"\n","    Function to calculate mean and standard deviation for categories in a DataFrame,\n","    and plot the top 10 most represented and most variable categories.\n","\n","    Args:\n","    df (pd.DataFrame): DataFrame containing the data.\n","    start_col (int): Column index from which to start the analysis.\n","    category_name (str): Name of the category (e.g., 'Bacteria', 'Genes') for labeling purposes.\n","    \"\"\"\n","\n","    # Calculate mean and standard deviation for each category\n","    category_mean = df.iloc[:, start_col:].mean()\n","    category_std = df.iloc[:, start_col:].std()\n","\n","    # Identify the most represented categories (highest mean)\n","    most_represented = category_mean.sort_values(ascending=False).head(10)\n","\n","    # Identify categories with the most variation (highest standard deviation)\n","    most_variable = category_std.sort_values(ascending=False).head(10)\n","\n","    # Plotting\n","    plt.figure(figsize=(20, 10))\n","    sns.set(style=\"whitegrid\", palette=\"muted\")\n","\n","    # Top 10 Most Represented Categories\n","    plt.subplot(1, 2, 1)\n","    ax1 = sns.barplot(x=most_represented.values, y=most_represented.index, palette=\"rocket\")\n","    plt.title(f'Top 10 Most Represented {category_name}', fontsize=18)\n","    plt.xlabel(f'Mean Relative Abundance for {category_name}', fontsize=14)\n","    plt.ylabel('')\n","    plt.xticks(fontsize=12)\n","    plt.yticks(fontsize=12)\n","\n","    # Remove axes spines\n","    ax1.spines['top'].set_visible(False)\n","    ax1.spines['right'].set_visible(False)\n","    ax1.spines['left'].set_visible(False)\n","    ax1.spines['bottom'].set_visible(False)\n","\n","    # Remove legend border if exists\n","    if ax1.get_legend():\n","        leg1 = ax1.get_legend()\n","        leg1.get_frame().set_linewidth(0.0)\n","\n","    # Top 10 Categories with Most Variation\n","    plt.subplot(1, 2, 2)\n","    ax2 = sns.barplot(x=most_variable.values, y=most_variable.index, palette=\"rocket\")\n","    plt.title(f'Top 10 {category_name} with Most Variation', fontsize=18)\n","    plt.xlabel(f'Standard Deviation of Relative Abundance for {category_name}', fontsize=14)\n","    plt.ylabel('')\n","    plt.xticks(fontsize=12)\n","    plt.yticks(fontsize=12)\n","\n","    # Remove axes spines\n","    ax2.spines['top'].set_visible(False)\n","    ax2.spines['right'].set_visible(False)\n","    ax2.spines['left'].set_visible(False)\n","    ax2.spines['bottom'].set_visible(False)\n","\n","    # Remove legend border if exists\n","    if ax2.get_legend():\n","        leg2 = ax2.get_legend()\n","        leg2.get_frame().set_linewidth(0.0)\n","\n","    plt.tight_layout()\n","    plt.show()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":1,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"YEOSFvPdwdC8"},"outputs":[],"source":["# @title plot species abundance over time\n","\n","def plot_species_abundance_over_time(df, cols_to_plot, title, type_, selection_mode='user'):\n","    \"\"\"\n","    Plots the abundance over time for the specified species for each patient in the dataframe in a single figure with subplots.\n","\n","    Args:\n","    df (pd.DataFrame): DataFrame containing microbial abundance data.\n","    cols_to_plot (list or None): List of species names to plot, or None if to be determined by selection_mode.\n","    title (str): Title for the plots and y-axis label.\n","    type_ (str): Label for the plot legend.\n","    selection_mode (str): Mode of selecting columns ('user', 'variation', 'mean'). 'user' expects cols_to_plot to be provided.\n","    \"\"\"\n","    unique_patients = df['patient_id'].unique()\n","    num_patients = len(unique_patients)\n","    cols = 2  # Number of columns for subplots\n","    rows = (num_patients // cols) + (num_patients % cols > 0)  # Calculate rows needed\n","\n","    fig, axs = plt.subplots(rows, cols, figsize=(20, 5 * rows))  # Adjust figsize as needed\n","    axs = axs.flatten()  # Flatten in case of 2D array of axes\n","\n","    for i, patient in enumerate(unique_patients):\n","        ax = axs[i]\n","        patient_data = df[df['patient_id'] == patient].copy()\n","        patient_data.sort_values(by='week', inplace=True)  # Ensure data is sorted by week\n","\n","        # Exclude non-abundance columns to focus on species data\n","        species_data = patient_data.drop(exclude_columns, axis=1)\n","\n","        # Determine columns to plot based on selection_mode if cols_to_plot is None or selection_mode is not 'user'\n","        if cols_to_plot is None or selection_mode != 'user':\n","            if selection_mode == 'variation':\n","                # Find the x species with the highest standard deviation\n","                cols_to_plot = species_data.std().sort_values(ascending=False).head(6).index.tolist()\n","            elif selection_mode == 'top1':\n","                top_species = set()  # Use a set to avoid duplicates\n","                for index, row in species_data.iterrows():\n","                    top_species.update(row.nlargest(1).index)\n","                cols_to_plot = list(top_species)  # Convert set to list\n","            elif selection_mode == 'mean':\n","                # Find the x species with the highest mean\n","                cols_to_plot = species_data.mean().sort_values(ascending=False).head(6).index.tolist()\n","\n","        # Plot the abundance over time for the specified species\n","        for species in cols_to_plot:\n","            if species in species_data.columns:\n","                sns.lineplot(data=patient_data, x='week', y=species, label=species, ax=ax)\n","\n","        # Annotate flare status changes (if applicable)\n","        flare_changes = patient_data.drop_duplicates('Flare_status', keep='first')[['week', 'Flare_status']]\n","        for _, (week, status) in flare_changes.iterrows():\n","            ax.axvline(x=week, color='grey', linestyle='--')\n","            ax.text(week, ax.get_ylim()[1], f' {status}', verticalalignment='top', fontsize=8)\n","\n","        # Setting x-axis ticks for every week, adjusting label rotation for clarity\n","        major_ticks = range(55)[::5]  # Get unique weeks\n","        ax.set_xticks(major_ticks)\n","        ax.set_xticklabels(major_ticks, rotation=45)\n","\n","        ax.set_title(f'{title} Over Time for Patient {patient}')\n","        ax.set_xlabel('Week')\n","        ax.set_ylabel(title)\n","        ax.legend(title=type_, bbox_to_anchor=(1.05, 1), loc='upper left')\n","        ax.grid(False)\n","\n","        # Remove axes spines\n","        ax.spines['top'].set_visible(False)\n","        ax.spines['right'].set_visible(False)\n","        ax.spines['left'].set_visible(True)\n","        ax.spines['bottom'].set_visible(True)\n","\n","    # Hide any unused subplots\n","    for j in range(i + 1, len(axs)):\n","        fig.delaxes(axs[j])\n","\n","    # Remove legend border\n","    for leg in ax.get_legend().get_frame().axes.figure.legends:\n","        leg.get_frame().set_linewidth(0.0)\n","\n","    plt.tight_layout()\n","    plt.show()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":1,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"IXnbj8AdwmYg","cellView":"form"},"outputs":[],"source":["# @title plot species abundance over time 2\n","\n","def plot_species_abundance_over_time_2(df, cols_to_plot, title, type_, poly_degree=3):\n","    \"\"\"\n","    Plots the abundance over time for the specified species for all patients in the dataframe,\n","    with separate plots for patients with and without a flare start, all in a single figure with subplots.\n","    Includes a trend line for each plot using polynomial regression.\n","\n","    Args:\n","    df (pd.DataFrame): DataFrame containing microbial abundance data.\n","    cols_to_plot (list): List of species names to plot.\n","    title (str): Title for the plots and y-axis label.\n","    type_ (str): Label for the plot legend.\n","    poly_degree (int): Degree of the polynomial for the trend line. Default is 3.\n","    \"\"\"\n","    if not cols_to_plot:\n","        raise ValueError(\"cols_to_plot must be provided when selection_mode is 'user'\")\n","\n","    num_species = len(cols_to_plot)\n","    cols = 2  # Number of columns for subplots (one for with flare, one for without)\n","    rows = num_species  # Each species will occupy one row with two subplots\n","\n","    fig, axs = plt.subplots(rows, cols, figsize=(20, 5 * rows))  # Adjust figsize as needed\n","\n","    # If only one species, axs might not be a 2D array, ensure it is\n","    if num_species == 1:\n","        axs = [axs]\n","\n","    # Determine patients where the first recorded week is a flare start\n","    df['first_week'] = df.groupby('patient_id')['week'].transform('min')  # Find first week for each patient\n","    patients_with_flare = df[df['Flare_start'] == 1]['patient_id'].unique()\n","    patients_without_flare = df[~df['patient_id'].isin(patients_with_flare)]['patient_id'].unique()\n","\n","    for i, species in enumerate(cols_to_plot):\n","        for j, (patients, title_suffix) in enumerate([(patients_with_flare, \"with Flare\"), (patients_without_flare, \"without Flare\")]):\n","            ax = axs[i][j]\n","\n","            all_weeks = []\n","            all_abundances = []\n","\n","            for patient in sorted(patients):  # Sort patients for consistent order\n","                patient_data = df[df['patient_id'] == patient].copy()\n","                patient_data.sort_values(by='week', inplace=True)  # Ensure data is sorted by week\n","\n","                if species in patient_data.columns:\n","                    sns.lineplot(data=patient_data, x='week', y=species, ax=ax, label=f'Patient {patient}')\n","\n","                    all_weeks.extend(patient_data['week'])\n","                    all_abundances.extend(patient_data[species])\n","\n","                if j == 0:  # Only for patients with flare\n","                    flare_starts = patient_data[patient_data['Flare_start'] == 1]\n","                    ax.scatter(flare_starts['week'], flare_starts[species],\n","                               color='red', s=10, zorder=5, label='Flare Start' if patient == patients_with_flare[0] else \"\")\n","\n","            # # Add trend line using polynomial regression\n","            # if all_weeks and all_abundances:\n","            #     z = np.polyfit(all_weeks, all_abundances, poly_degree)\n","            #     p = np.poly1d(z)\n","            #     x_range = np.linspace(min(all_weeks), max(all_weeks), 100)\n","            #     ax.plot(x_range, p(x_range),  label='Trend line', color= \"red\", linestyle =\"--\")\n","\n","            ax.set_title(f'{title} ({species}) Over Time for Patients {title_suffix}')\n","            ax.set_xlabel('Week')\n","            ax.set_ylabel(title)\n","\n","            # Remove axes spines\n","            ax.spines['top'].set_visible(False)\n","            ax.spines['right'].set_visible(False)\n","            ax.spines['left'].set_visible(True)\n","            ax.spines['bottom'].set_visible(True)\n","\n","            # Create legend with no box and sorted alphabetically\n","            handles, labels = ax.get_legend_handles_labels()\n","            by_label = dict(zip(labels, handles))\n","            sorted_labels = sorted(by_label.keys())\n","            ax.legend(title=type_, bbox_to_anchor=(1.05, 1), loc='upper left',\n","                      handles=[by_label[l] for l in sorted_labels],\n","                      labels=sorted_labels, frameon=False)\n","\n","            major_ticks = range(0, max(all_weeks) + 1, 5)  # Get unique weeks\n","            ax.set_xticks(major_ticks)\n","            ax.set_xticklabels(major_ticks, rotation=45)\n","            ax.grid(False)\n","\n","    plt.tight_layout()\n","    plt.show()"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":1,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"o07JFPZGwuYT"},"outputs":[],"source":["# @title segment_time_series\n","def segment_time_series(df, length=6):\n","    # Sort by patient and week to ensure chronological order\n","    df.sort_values(by=['patient_id', 'week'], inplace=True)\n","\n","    # Create 'ts' column initialized to NaN\n","    df['ts'] = np.nan\n","\n","    # Process each patient individually\n","    for patient in df['patient_id'].unique():\n","        patient_data = df[df['patient_id'] == patient]\n","\n","        # Check for the presence of a flare start\n","        flare_start_indices = patient_data[patient_data['Flare_start'] == 1].index.tolist()\n","\n","        # Initial segment number\n","        ts_value = 1\n","\n","        if flare_start_indices:\n","            # There is at least one flare start\n","            flare_start_index = flare_start_indices[0]  # Assume first flare start is the one to consider\n","            # Calculate the position of the first flare start\n","            flare_position = patient_data.index.get_loc(flare_start_index)\n","\n","            # Calculate the length of the first segment to align flare start with a new segment start\n","            if flare_position % length == 0:\n","                first_segment_length = 0  # Flare start is already aligned with a new segment\n","            else:\n","                first_segment_length = flare_position % length\n","\n","            # Assign time series IDs\n","            # Segment before flare start if it has any length\n","            if first_segment_length > 0:\n","                df.loc[patient_data.index[:first_segment_length], 'ts'] = ts_value\n","                ts_value += 1\n","\n","            # Remaining data after the initial adjustment\n","            for i in range(first_segment_length, len(patient_data), length):\n","                df.loc[patient_data.index[i:i+length], 'ts'] = ts_value\n","                ts_value += 1\n","\n","        else:\n","            # No flare, segment normally every x points\n","            for i in range(0, len(patient_data), length):\n","                df.loc[patient_data.index[i:i+length], 'ts'] = ts_value\n","                ts_value += 1\n","\n","    # Convert 'ts' column to integer\n","    df['ts'] = df['ts'].astype(int)\n","\n","    # Create a new column combining 'ts' and 'patient_id'\n","    df['ts_patient_id'] = df['patient_id'] + '_TS' + df['ts'].astype(str)\n","\n","    return df\n"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":1,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"uGODlpm6nXhe"},"outputs":[],"source":["# @title Augmented Dickey-Fuller (ADF) test\n","def analyze_time_series(data, significance_level=0.05, mht=\"bonferroni\", id_col=\"ts_patient_id\"):\n","    #flare_patients = data[data['Flare_status']==\"During_flare\"]['patient_id'].unique()\n","    target_ids = data[data[\"RBF\"]==True]['ts_patient_id'].unique()\n","    print(f'Target ids : {target_ids}')\n","\n","    # Suppress runtime warnings from statsmodels or other calculations\n","    warnings.filterwarnings('ignore', category=RuntimeWarning)\n","\n","    filtered_df = data.groupby('ts_patient_id').filter(lambda x: len(x) > 4)\n","    print(exclude_columns)\n","\n","    # Try to concatenate directly\n","    col_to_exclude = exclude_columns.tolist() + ['ts', 'RBF', 'ts_patient_id']\n","\n","\n","    # Define columns to analyze (excluding metadata columns)\n","    analyze_columns = [col for col in filtered_df.columns if col not in col_to_exclude]\n","\n","    # Prepare to store results\n","    adf_results = []\n","\n","    # Perform ADF tests\n","    for id in filtered_df[id_col].unique():\n","        patient_data = filtered_df[filtered_df[id_col] == id]\n","        for column in analyze_columns:\n","            time_series = patient_data[column]\n","            if time_series.notna().all() and len(time_series) > 1 and time_series.var() > 0:\n","                try:\n","                    result = adfuller(time_series)\n","                    adf_results.append({\n","                        'id': id,\n","                        'variable': column,\n","                        'ADF Statistic': result[0],\n","                        'p-value': result[1],\n","                        'Used lag': result[2],\n","                        'Number of observations': result[3],\n","                        'Critical values': result[4],\n","                        'IC used': result[5]\n","                    })\n","                except Exception as e:\n","                    adf_results.append({\n","                        'id': id,\n","                        'variable': column,\n","                        'Error': str(e)\n","                    })\n","            else:\n","                adf_results.append({\n","                    'id': id,\n","                    'variable': column,\n","                    'Error': 'Insufficient data or variance'\n","                })\n","\n","    # Convert results to DataFrame\n","    results_df = pd.DataFrame(adf_results)\n","    results_df['p-value'] = pd.to_numeric(results_df['p-value'], errors='coerce')\n","\n","    # Apply multiple testing corrections\n","    _, bh_adjusted, _, _ = multipletests(results_df['p-value'].dropna(), method='fdr_bh')\n","    _, bonf_adjusted, _, _ = multipletests(results_df['p-value'].dropna(), method='bonferroni')\n","\n","    results_df['bh_adjusted_p-value'] = pd.Series(bh_adjusted, index=results_df['p-value'].dropna().index)\n","    results_df['bonferroni_adjusted_p-value'] = pd.Series(bonf_adjusted, index=results_df['p-value'].dropna().index)\n","\n","    # Filter significant results\n","    if mht == \"bonferroni\":\n","      significant_results = results_df[results_df['bonferroni_adjusted_p-value'] < significance_level]\n","    elif mht == \"bh\" :\n","      significant_results = results_df[results_df['bh_adjusted_p-value'] < significance_level]\n","\n","    # Print results\n","    significant_counts = significant_results.groupby('id').size()\n","    print(\"Summary of significant adjusted p-value counts by sub-timeseries:\")\n","    print(significant_counts.sort_values(ascending=False))\n","\n","    for id, count in significant_counts.items():\n","        print(f\"Patient ID: {id}\")\n","        print(f\"Number of variables with significant adjusted p-values: {count}\")\n","        patient_vars = significant_results[significant_results['id'] == id]\n","        #for index, row in patient_vars.iterrows():\n","         # if mht == \"bonferroni\":\n","          #  print(f\"  Variable: {row['variable']}, Adjusted p-value: {row['bonferroni_adjusted_p-value']:.5f}\")\n","          #elif mht == \"bh\" :\n","           # print(f\"  Variable: {row['variable']}, Adjusted p-value: {row['bh_adjusted_p-value']:.5f}\")\n","        #print(\"\\n\")  # Add a new line for readability\n","\n","\n","    #target_ids= ['TR_2101_TS4', 'TR_2102_TS2']\n","\n","    # Find and print variables significant across specific target patient IDs\n","    target_significant = significant_results[significant_results['id'].isin(target_ids)]\n","    common_significant_vars = target_significant.groupby('variable').filter(lambda x: len(x) >= 0.5*len(target_ids))\n","\n","    if not common_significant_vars.empty:\n","        print(\"\\nVariables significant across at least 50% of target patient IDs:\")\n","        for variable in common_significant_vars['variable'].unique():\n","            print(f\"Variable: {variable}\")\n","            relevant_patients = significant_results[significant_results['variable'] == variable]\n","            other_patients = relevant_patients[~relevant_patients['id'].isin(target_ids)]\n","            print(f\"Significant in target IDs: {', '.join(relevant_patients['id'].unique())}\")\n","            if not other_patients.empty:\n","                print(f\"Also significant in other {len(other_patients['id'].unique())} portions\")\n","            else:\n","                print(\"Not significant in any other patient IDs.\")\n","    else:\n","        print(\"No variables are significant across all specified target patient IDs.\")\n","\n","    print(\"\\n\")  # Add a new line for readability"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":1,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"9Y9M-La-akh-"},"outputs":[],"source":["# @title smooth_patient_data\n","def smooth_patient_data(df, value_columns, patient_column='patient_id', week_column='week'):\n","    \"\"\"\n","    Smooths the values of the specified columns in the dataframe using a rolling window average.\n","\n","    Args:\n","    df (pd.DataFrame): DataFrame containing the data to be smoothed.\n","    value_columns (list): List of columns to smooth.\n","    patient_column (str): Column name for patient ID.\n","    week_column (str): Column name for week number.\n","\n","    Returns:\n","    pd.DataFrame: Smoothed DataFrame.\n","    \"\"\"\n","    df = df.copy()\n","    smoothed_dfs = []\n","\n","    for patient in df[patient_column].unique():\n","        patient_data = df[df[patient_column] == patient].copy()\n","        patient_data = patient_data.sort_values(by=week_column)\n","\n","        for col in value_columns:\n","            smoothed_values = patient_data[col].rolling(window=3, min_periods=1, center=True).mean()\n","            # Handle first and last weeks separately\n","            if len(smoothed_values) > 1:\n","                smoothed_values.iloc[0] = patient_data[col].iloc[:2].mean()\n","                smoothed_values.iloc[-1] = patient_data[col].iloc[-2:].mean()\n","            patient_data[col] = smoothed_values\n","\n","        smoothed_dfs.append(patient_data)\n","\n","    smoothed_df = pd.concat(smoothed_dfs)\n","    return smoothed_df\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3882,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"44AqjPY5dWlB"},"outputs":[],"source":["# @title\n","species_columns = [col for col in metaphlan_df.columns if col.startswith('s__')]\n","\n","# Group by 'patient_id' and calculate the mean of each species\n","mean_abundances = metaphlan_df.groupby('patient_id')[species_columns].mean()\n","\n","\n","# and print the results for each patient\n","for patient_id, abundances in mean_abundances.iterrows():\n","    print(f\"Patient {patient_id} has the following species with an average relative abundance above 0.07:\")\n","    filtered_species = abundances[abundances > 0.07 ]\n","    if not filtered_species.empty:\n","        print(filtered_species)\n","    else:\n","        print(\"No species above the threshold.\")\n","    print()"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":3881,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"16h6Q08CZzPt"},"outputs":[],"source":["# @title Plot Microbial abundance profiles\n","combinations = []\n","\n","# Loop until we have 10 good combinations\n","while len(combinations) < 16:\n","    # Randomly select a patient_id and column name\n","    patient_id = metaphlan_df['patient_id'].sample().values[0]\n","    column_name = metaphlan_df.columns[np.random.randint(1, len(metaphlan_df.columns))]\n","\n","    # Check if the values for the selected patient_id and column are not all zeros\n","    if not all(metaphlan_df[metaphlan_df['patient_id'] == patient_id][column_name] == 0):\n","        # Add the combination to the list\n","        combinations.append((patient_id, column_name))\n","\n","# Print the list of good combinations\n","print(combinations)\n","\n","\n","# Define the number of rows and columns for the subplot grid\n","nrows = 4\n","ncols = 4\n","\n","# Create the figure and subplots\n","fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(16, 12))\n","sns.set(style=\"whitegrid\")\n","# Iterate through the list of combinations\n","for i, (patient_id, column_name) in enumerate(combinations):\n","    # Select the current subplot\n","    ax = axes.flat[i]\n","\n","    # Plot the histogram of values\n","    sns.histplot(data=metaphlan_df[metaphlan_df['patient_id'] == patient_id][column_name], ax=ax, bins = 30, color='skyblue', kde=True)\n","\n","    # Set the title of the subplot\n","    ax.set_title(f'{patient_id} - {column_name}')\n","\n","\n","# Adjust the spacing between subplots\n","plt.tight_layout()\n","\n","# Show the plot\n","plt.show()"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3880,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"Zq05B6E-CCQ1"},"outputs":[],"source":["exclude_columns = metaphlan_df.columns.difference(species_columns)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3880,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"5osQdEdMom5O"},"outputs":[],"source":["# # @title CLR transformation for Microbial Composition\n","# def clr_transform(df, metadata_columns):\n","#     species_data = df.drop(columns=metadata_columns)\n","#     species_data += 1e-5\n","#     geometric_means = gmean(species_data, axis=1)\n","#     clr_transformed_data = np.log(species_data.divide(geometric_means, axis=0))\n","#     clr_transformed_df = pd.concat([df[metadata_columns], clr_transformed_data], axis=1)\n","#     return clr_transformed_df\n","\n","# metaphlan_df = clr_transform(metaphlan_df, exclude_columns)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3878,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"tNDngLFqZzoq"},"outputs":[],"source":["# @title Plot Microbial Abundance Profiles after CLR transformation\n","# Create the figure and subplots\n","fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(16, 12))\n","sns.set(style=\"whitegrid\")\n","\n","# Iterate through the list of combinations\n","for i, (patient_id, column_name) in enumerate(combinations):\n","    # Select the current subplot\n","    ax = axes.flat[i]\n","\n","    # Plot the histogram of values\n","    sns.histplot(data=metaphlan_df[metaphlan_df['patient_id'] == patient_id][column_name], ax=ax, bins = 30,  color='skyblue', kde=True)\n","\n","    # Set the title of the subplot\n","    ax.set_title(f'{patient_id} - {column_name}')\n","\n","\n","# Adjust the spacing between subplots\n","plt.tight_layout()\n","\n","# Show the plot\n","plt.show()"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3878,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"oqDNhqCGkXLz"},"outputs":[],"source":["# @title Adding a Flare_start column\n","metaphlan_df = flag_first_flare_weeks(metaphlan_df)\n","gene_count_df = flag_first_flare_weeks(gene_count_df)\n","exclude_columns+= ['Flare_start']"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3876,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"vlJ94hxmMQEc"},"outputs":[],"source":["analyze_representation_and_variation(metaphlan_df, 5, 'Bacteria')\n","analyze_representation_and_variation(gene_count_df, 5, 'Genes')"]},{"cell_type":"markdown","metadata":{"id":"SMUQnI4Cxv3K"},"source":["Patient 2102 had a flair from week 16 to 24\n","Patient 2101 had a flair from week 25 to 43\n","Patient 2103 had a flair from week 27 to 39"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":3876,"status":"aborted","timestamp":1723144946542,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"GKwqMNRvaTd2"},"outputs":[],"source":["# @title PCOA functions\n","def perform_pcoa(data, columns_to_exclude, patient_id_column='patient_id', week_column='week', flare_status_column='Flare_status'):\n","    \"\"\"\n","    Performs PCoA on the given data, excluding specified columns.\n","\n","    Parameters:\n","    data (pd.DataFrame): The dataframe containing the data.\n","    columns_to_exclude (list of str): List of columns to exclude from the analysis.\n","    patient_id_column (str): The column name for patient ID. Default is 'patient_id'.\n","    week_column (str): The column name for week. Default is 'week'.\n","    flare_status_column (str): The column name for flare status. Default is 'Flare_status'.\n","\n","    Returns:\n","    pd.DataFrame: A dataframe containing the PCoA results with patient ID, week, and flare status.\n","    \"\"\"\n","    data_filtered = data.drop(columns=columns_to_exclude)\n","\n","    pca = PCA(n_components=2)\n","    pcoa_results = pca.fit_transform(data_filtered)\n","\n","    pcoa_df = pd.DataFrame(pcoa_results, columns=['PC1', 'PC2'])\n","    pcoa_df[patient_id_column] = data[patient_id_column].values\n","    pcoa_df[week_column] = data[week_column].values\n","    pcoa_df[flare_status_column] = data[flare_status_column].values\n","\n","    return pcoa_df\n","\n","\n","def create_combined_pcoa_plots(microbial_data, gene_expression_data, patient_ids):\n","    \"\"\"\n","    Creates 3D PCoA plots for microbial abundances and gene expression for all patients in a single figure,\n","    with rows of 4 plots.\n","\n","    Parameters:\n","    microbial_data (pd.DataFrame): The dataframe containing microbial abundances.\n","    gene_expression_data (pd.DataFrame): The dataframe containing gene expression data.\n","    patient_ids (list of str): List of patient IDs to create plots for.\n","\n","    Returns:\n","    None\n","    \"\"\"\n","    n_patients = len(patient_ids)\n","    n_rows = (n_patients + 1) // 2  # Ceil division to get number of rows\n","    fig = plt.figure(figsize=(20, 3*n_rows))\n","\n","    for i, patient_id in enumerate(patient_ids):\n","        # Filter data for the current patient\n","        microbial_patient_data = microbial_data[microbial_data['patient_id'] == patient_id]\n","        gene_expression_patient_data = gene_expression_data[gene_expression_data['patient_id'] == patient_id]\n","\n","        microbial_pcoa_df = perform_pcoa(microbial_patient_data, exclude_columns)\n","        gene_expression_pcoa_df = perform_pcoa(gene_expression_patient_data, exclude_columns)\n","\n","        # Plot 3D PCoA for microbial data\n","        ax1 = fig.add_subplot(n_rows, 4, 2*i + 1, projection='3d')\n","        plot_pcoa_3d_subplot(microbial_pcoa_df, patient_id, 'Microbial Abundances', ax1)\n","\n","        # Plot 3D PCoA for gene expression data\n","        ax2 = fig.add_subplot(n_rows, 4, 2*i + 2, projection='3d')\n","        plot_pcoa_3d_subplot(gene_expression_pcoa_df, patient_id, 'Gene Expression', ax2)\n","\n","    plt.tight_layout()\n","    plt.show()\n","\n","def plot_pcoa_3d_subplot(pcoa_df, patient_id, title, ax, flare_status_column='Flare_status'):\n","    \"\"\"\n","    Plots 3D PCoA results on a given subplot.\n","\n","    Parameters:\n","    pcoa_df (pd.DataFrame): The dataframe containing the PCoA results.\n","    patient_id (str): The patient ID for the plot title.\n","    title (str): The title for the plot.\n","    ax (matplotlib.axes.Axes): The subplot to plot on.\n","    flare_status_column (str): The column name for flare status. Default is 'Flare_status'.\n","\n","    Returns:\n","    None\n","    \"\"\"\n","    flare_status_mapping = {\n","        'Pre_flare': '#e78ac3',\n","        'Pre_flare_two': '#e78ac3',\n","        'During_flare': '#ff7f00',\n","        'During_flare_2': '#ff7f00',\n","        'Post_flare': '#8da0cb',\n","        'Post_flare_2': '#8da0cb',\n","        'No_flare': '#66c2a5',\n","        'Before_infusion': 'pink',\n","        'After_infusion': 'orange'\n","    }\n","\n","    pcoa_df['Flare_status_color'] = pcoa_df[flare_status_column].map(flare_status_mapping)\n","\n","    scatter = ax.scatter(pcoa_df['week'], pcoa_df['PC1'], pcoa_df['PC2'], c=pcoa_df['Flare_status_color'], s=20, alpha=0.7)\n","\n","    ax.set_xlabel('Week')\n","    ax.set_ylabel('PC1')\n","    ax.set_zlabel('PC2')\n","    ax.set_title(f'Patient {patient_id} - {title}')\n","\n","    # Create a custom legend for elements present in the plot\n","    unique_flare_statuses = pcoa_df[flare_status_column].unique()\n","    legend_handles = [mpatches.Patch(color=flare_status_mapping[status], label=status) for status in unique_flare_statuses if status in flare_status_mapping]\n","    ax.legend(handles=legend_handles, title='Flare Status', bbox_to_anchor=(1.05, 1), loc='upper left')\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":0,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"ddC0ect3Crfi"},"outputs":[],"source":["exclude_columns = metaphlan_df.columns.difference(species_columns)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4169,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"yYuvTg8QaY3z"},"outputs":[],"source":["# Use the function\n","patient_ids = metaphlan_df['patient_id'].unique()\n","create_combined_pcoa_plots(metaphlan_df, gene_count_df, patient_ids)"]},{"cell_type":"code","source":["\n","extended_colors = [\n","    '#66c2a5',  # Teal\n","    '#ff7f00',  # Orange\n","    '#8da0cb',  # Light Blue\n","    '#e78ac3',  # Pink\n","    '#a6d854',  # Light Green\n","    '#e5c494',  # Tan\n","    '#b3b3b3',  # Gray\n","    '#80b1d3',  # Sky Blue\n","    '#bebada',  # Lavender\n","    '#fb8072',  # Light Red\n","    '#7fc97f',  # Medium Green\n","    '#d9d9d9',  # Light Gray\n","    '#ffed6f',  # Pale Yellow\n","    '#b15928',  # Brown\n","    '#fccde5',  # Pale Pink\n","    '#8dd3c7',  # Pale Teal\n","    '#ffffb3',  # Very Pale Yellow\n","    '#bc80bd'   # Purple\n","]\n","\n","# Set default color cycle\n","plt.rcParams['axes.prop_cycle'] = plt.cycler(color=extended_colors)\n","\n","# Set other default styles\n","#plt.rcParams['figure.facecolor'] = '#f0f0f0'\n","plt.rcParams['axes.facecolor'] = 'white'\n","plt.rcParams['grid.color'] = '#cccccc'\n","plt.rcParams['grid.linestyle'] = '--'\n","plt.rcParams['grid.alpha'] = 0.5\n","\n","# Adjust grid frequency\n","plt.rcParams['axes.grid'] = False\n","plt.rcParams['axes.grid.which'] = 'major'\n","plt.rcParams['axes.grid.axis'] = 'both'\n","\n","\n","# Set default line width\n","plt.rcParams['lines.linewidth'] = 1"],"metadata":{"id":"DQzxYbOwQ4U1","executionInfo":{"status":"aborted","timestamp":1723144946836,"user_tz":240,"elapsed":4169,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4167,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"l9_ENoAOgmmm"},"outputs":[],"source":["plot_species_abundance_over_time(metaphlan_df, None, 'Microbial Abundance', 'Bacteria', 'mean')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4166,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"D6FlHhIkCfkx"},"outputs":[],"source":["plot_species_abundance_over_time(metaphlan_df, None, 'Microbial Abundance', 'Bacteria', 'variation')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4165,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"kg9ltvJ8Zokd"},"outputs":[],"source":["plot_species_abundance_over_time(gene_count_df, None, 'Gene Expression', 'Genes', 'mean')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4164,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"ObcPZXX5CmhN"},"outputs":[],"source":["plot_species_abundance_over_time(gene_count_df, None, 'Gene Expression', 'Genes', 'variation')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4163,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"H7hwPfO6CtH5"},"outputs":[],"source":["x = ['s__Blautia_SGB4805', 's__Blautia_SGB4815', 's__Blautia_SGB4831',\n","       's__Blautia_argi', 's__Blautia_caecimuris', 's__Blautia_faecicola',\n","       's__Blautia_faecis', 's__Blautia_glucerasea', 's__Blautia_hansenii',\n","       's__Blautia_hydrogenotrophica', 's__Blautia_massiliensis',\n","       's__Blautia_obeum', 's__Blautia_producta', 's__Blautia_sp_AF19_10LB',\n","       's__Blautia_sp_An249', 's__Blautia_sp_An81', 's__Blautia_sp_MSK_20_85',\n","       's__Blautia_sp_MSK_21_1', 's__Blautia_sp_OF03_15BH',\n","       's__Blautia_stercoris', 's__Blautia_wexlerae']+['s__Phocaeicola_dorei','s__Bacteroides_uniformis','s__Faecalibacterium_prausnitzii' , 's__Bifidobacterium_pullorum',\n","  's__Bacteroides_ovatus', 's__Clostridia_bacterium', 's__Eubacterium_rectale', 's__Blautia_sp_MSK_21_1',\n","  's__GGB4569_SGB6310', 's__Blautia_wexlerae']\n","\n","x.sort()\n","x = list(dict.fromkeys(x))\n","x"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":4163,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"yd8898jFfFbK"},"outputs":[],"source":["# @title Remove dominant species\n","# # Step 1: Compute mean relative abundances for each species per patient\n","# species_columns = [col for col in metaphlan_df.columns if col.startswith('s__')]\n","# mean_abundances = metaphlan_df.groupby('patient_id')[species_columns].mean()\n","\n","# # Step 2: Identify species to zero out\n","# species_to_zero = mean_abundances.apply(lambda x: x > 0.07)\n","\n","# # Step 3: Set values to zero for identified species\n","# for patient in species_to_zero.index:\n","#     for species in species_columns:\n","#         if species_to_zero.at[patient, species]:\n","#             metaphlan_df.loc[metaphlan_df['patient_id'] == patient, species] = 0\n","\n","# # Step 4: Normalize the DataFrame so that each row sums to 1\n","# metaphlan_df[species_columns] = metaphlan_df[species_columns].div(metaphlan_df[species_columns].sum(axis=1), axis=0)"]},{"cell_type":"markdown","metadata":{"id":"Ol2Zz4W1nCZ7"},"source":["microbes to investigate"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":4163,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"H9N4YQotG3vV"},"outputs":[],"source":["# @title assign_rbf\n","def assign_rbf(df, samples_before_flare, strict=False):\n","    # Initial setup\n","    df = df.sort_values(by=['patient_id', 'week'])\n","    df['RBF'] = 0  # Initialize the RBF column\n","\n","    # Identify patients and all weeks where flares start\n","    flare_starts = df[df['Flare_start'] == 1].groupby('patient_id')['week'].apply(list)\n","\n","    for patient, flare_weeks in flare_starts.items():\n","        # Handle each flare start week\n","        for flare_week in flare_weeks:\n","            # Get all weeks for the current patient\n","            patient_weeks = df.loc[df['patient_id'] == patient, 'week']\n","\n","            if strict:\n","                # Find indexes for weeks strictly before this flare start and within the limit\n","                weeks_before_flare = patient_weeks[(patient_weeks < flare_week) & (patient_weeks >= flare_week - samples_before_flare)]\n","            else:\n","                # Find indexes for weeks before this flare start\n","                weeks_before_flare = patient_weeks[patient_weeks < flare_week].nlargest(samples_before_flare)\n","\n","\n","            # Set RBF to 1 for these weeks\n","            df.loc[(df['patient_id'] == patient) & (df['week'].isin(weeks_before_flare)), 'RBF'] = 1\n","\n","    return df"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4162,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"16O3XF_iG6Uh"},"outputs":[],"source":["metaphlan_df = assign_rbf(metaphlan_df, 6, strict=False)\n","gene_count_df = assign_rbf(gene_count_df, 6, strict=False)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4162,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"Dqw-ue1amNW8"},"outputs":[],"source":["# @title Segment time series\n","# Apply the function to both dataframes\n","metaphlan_df = segment_time_series(metaphlan_df, 6)\n","gene_count_df = segment_time_series(gene_count_df, 6)\n","\n","# Check the results\n","print(metaphlan_df[['patient_id', 'week', 'Flare_status', 'Flare_start', 'ts', 'ts_patient_id', 'RBF']])\n","print(gene_count_df[['patient_id', 'week', 'Flare_status', 'Flare_start', 'ts', 'ts_patient_id','RBF']])"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4162,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"zOW5r292zzT7"},"outputs":[],"source":["analyze_time_series(metaphlan_df, 0.1, \"bh\")"]},{"cell_type":"code","source":["# @title copy paste output here to process it\n","import re\n","\n","output = \"\"\"\n","Patient ID: TR_2101_TS2\n","Number of variables with significant adjusted p-values: 44\n","Patient ID: TR_2101_TS3\n","Number of variables with significant adjusted p-values: 61\n","Patient ID: TR_2101_TS4\n","Number of variables with significant adjusted p-values: 52\n","Patient ID: TR_2101_TS5\n","Number of variables with significant adjusted p-values: 51\n","Patient ID: TR_2101_TS6\n","Number of variables with significant adjusted p-values: 40\n","Patient ID: TR_2101_TS7\n","Number of variables with significant adjusted p-values: 33\n","Patient ID: TR_2101_TS8\n","Number of variables with significant adjusted p-values: 50\n","Patient ID: TR_2102_TS1\n","Number of variables with significant adjusted p-values: 56\n","Patient ID: TR_2102_TS2\n","Number of variables with significant adjusted p-values: 58\n","Patient ID: TR_2102_TS3\n","Number of variables with significant adjusted p-values: 69\n","Patient ID: TR_2102_TS4\n","Number of variables with significant adjusted p-values: 62\n","Patient ID: TR_2102_TS5\n","Number of variables with significant adjusted p-values: 75\n","Patient ID: TR_2102_TS6\n","Number of variables with significant adjusted p-values: 37\n","Patient ID: TR_2102_TS7\n","Number of variables with significant adjusted p-values: 62\n","Patient ID: TR_2103_TS1\n","Number of variables with significant adjusted p-values: 83\n","Patient ID: TR_2103_TS2\n","Number of variables with significant adjusted p-values: 90\n","Patient ID: TR_2103_TS3\n","Number of variables with significant adjusted p-values: 48\n","Patient ID: TR_2103_TS4\n","Number of variables with significant adjusted p-values: 45\n","Patient ID: TR_2103_TS5\n","Number of variables with significant adjusted p-values: 26\n","Patient ID: TR_2103_TS6\n","Number of variables with significant adjusted p-values: 49\n","Patient ID: TR_2103_TS7\n","Number of variables with significant adjusted p-values: 42\n","Patient ID: TR_2104_TS1\n","Number of variables with significant adjusted p-values: 55\n","Patient ID: TR_2104_TS2\n","Number of variables with significant adjusted p-values: 127\n","Patient ID: TR_2104_TS3\n","Number of variables with significant adjusted p-values: 127\n","Patient ID: TR_2104_TS4\n","Number of variables with significant adjusted p-values: 75\n","Patient ID: TR_2104_TS5\n","Number of variables with significant adjusted p-values: 81\n","Patient ID: TR_2104_TS6\n","Number of variables with significant adjusted p-values: 62\n","Patient ID: TR_2104_TS7\n","Number of variables with significant adjusted p-values: 75\n","Patient ID: TR_2104_TS8\n","Number of variables with significant adjusted p-values: 63\n","Patient ID: TR_2105_TS1\n","Number of variables with significant adjusted p-values: 61\n","Patient ID: TR_2105_TS2\n","Number of variables with significant adjusted p-values: 50\n","Patient ID: TR_2105_TS3\n","Number of variables with significant adjusted p-values: 75\n","Patient ID: TR_2105_TS4\n","Number of variables with significant adjusted p-values: 54\n","Patient ID: TR_2105_TS5\n","Number of variables with significant adjusted p-values: 53\n","Patient ID: TR_2105_TS6\n","Number of variables with significant adjusted p-values: 102\n","Patient ID: TR_2105_TS7\n","Number of variables with significant adjusted p-values: 83\n","Patient ID: TR_2105_TS8\n","Number of variables with significant adjusted p-values: 71\n","Patient ID: TR_2106_TS1\n","Number of variables with significant adjusted p-values: 40\n","Patient ID: TR_2106_TS2\n","Number of variables with significant adjusted p-values: 58\n","Patient ID: TR_2106_TS3\n","Number of variables with significant adjusted p-values: 57\n","Patient ID: TR_2106_TS4\n","Number of variables with significant adjusted p-values: 49\n","Patient ID: TR_2106_TS5\n","Number of variables with significant adjusted p-values: 57\n","Patient ID: TR_2106_TS6\n","Number of variables with significant adjusted p-values: 47\n","Patient ID: TR_2106_TS7\n","Number of variables with significant adjusted p-values: 83\n","Patient ID: TR_2106_TS8\n","Number of variables with significant adjusted p-values: 47\n","Patient ID: TR_2107_TS2\n","Number of variables with significant adjusted p-values: 66\n","Patient ID: TR_2107_TS3\n","Number of variables with significant adjusted p-values: 79\n","Patient ID: TR_2107_TS4\n","Number of variables with significant adjusted p-values: 46\n","Patient ID: TR_2107_TS5\n","Number of variables with significant adjusted p-values: 63\n","Patient ID: TR_2107_TS6\n","Number of variables with significant adjusted p-values: 83\n","Patient ID: TR_2107_TS7\n","Number of variables with significant adjusted p-values: 77\n","Patient ID: TR_2108_TS1\n","Number of variables with significant adjusted p-values: 44\n","Patient ID: TR_2108_TS2\n","Number of variables with significant adjusted p-values: 48\n","Patient ID: TR_2108_TS3\n","Number of variables with significant adjusted p-values: 52\n","Patient ID: TR_2108_TS4\n","Number of variables with significant adjusted p-values: 64\n","Patient ID: TR_2201_TS1\n","Number of variables with significant adjusted p-values: 56\n","Patient ID: TR_2201_TS2\n","Number of variables with significant adjusted p-values: 69\n","Patient ID: TR_2201_TS3\n","Number of variables with significant adjusted p-values: 74\n","Patient ID: TR_2201_TS4\n","Number of variables with significant adjusted p-values: 73\n","Patient ID: TR_2201_TS5\n","Number of variables with significant adjusted p-values: 49\n","Patient ID: TR_2201_TS6\n","Number of variables with significant adjusted p-values: 49\n","Patient ID: TR_2201_TS7\n","Number of variables with significant adjusted p-values: 69\n","Patient ID: TR_2202_TS1\n","Number of variables with significant adjusted p-values: 39\n","Patient ID: TR_2202_TS2\n","Number of variables with significant adjusted p-values: 36\n","Patient ID: TR_2202_TS3\n","Number of variables with significant adjusted p-values: 22\n","Patient ID: TR_2202_TS4\n","Number of variables with significant adjusted p-values: 40\n","Patient ID: TR_2202_TS5\n","Number of variables with significant adjusted p-values: 51\n","Patient ID: TR_2202_TS6\n","Number of variables with significant adjusted p-values: 38\n","Patient ID: TR_2202_TS7\n","Number of variables with significant adjusted p-values: 31\n","Patient ID: TR_2202_TS8\n","Number of variables with significant adjusted p-values: 36\n","Patient ID: TR_2203_TS1\n","Number of variables with significant adjusted p-values: 33\n","Patient ID: TR_2203_TS2\n","Number of variables with significant adjusted p-values: 49\n","Patient ID: TR_2203_TS3\n","Number of variables with significant adjusted p-values: 58\n","Patient ID: TR_2203_TS4\n","Number of variables with significant adjusted p-values: 27\n","Patient ID: TR_2205_TS1\n","Number of variables with significant adjusted p-values: 99\n","\"\"\"\n"],"metadata":{"cellView":"form","id":"9YdpTAWY8EEb","executionInfo":{"status":"aborted","timestamp":1723144946836,"user_tz":240,"elapsed":4162,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["patients  = []\n","ts  = []\n","significant_p_values = []\n","\n","# Use regular expressions to extract the information\n","pattern = re.compile(r\"Patient ID: TR_(\\d+)_TS(\\d+)\\s+Number of variables with significant adjusted p-values: (\\d+)\")\n","\n","matches = pattern.findall(output)\n","\n","for match in matches:\n","    patients_, ts_, significant_p_values_ = match\n","    patients.append(int(patients_))\n","    ts.append(int(ts_))\n","    significant_p_values.append(int(significant_p_values_))"],"metadata":{"id":"Il_vru-Y8ehR","executionInfo":{"status":"aborted","timestamp":1723144946836,"user_tz":240,"elapsed":4161,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["highlight_set = set(metaphlan_df[metaphlan_df[\"RBF\"]==True]['ts_patient_id'].unique())\n","\n","# Generate combinations of Patient and TS\n","combinations = ['TR_' + str(p) + '_TS' + str(t) for p, t in zip(patients, ts)]\n","\n","# Get unique patients\n","unique_patients = sorted(set(patients))\n","\n","# Calculate the number of rows and columns for subplots\n","n = len(unique_patients)\n","cols = 3\n","rows = math.ceil(n / cols)\n","\n","# Create the plot\n","fig, axes = plt.subplots(rows, cols, figsize=(15, 5*rows))\n","\n","for i, patient in enumerate(unique_patients):\n","    row = i // cols\n","    col = i % cols\n","    ax = axes[row, col] if rows > 1 else axes[col]\n","\n","    # Filter data for the current patient\n","    patient_indices = [j for j, p in enumerate(patients) if p == patient]\n","    patient_ts = [ts[j] for j in patient_indices]\n","    patient_significant_p_values = [significant_p_values[j] for j in patient_indices]\n","    patient_combinations = [combinations[j] for j in patient_indices]\n","\n","    # Plot the data\n","    ax.plot(patient_ts, patient_significant_p_values, label=f'Patient {patient}')\n","\n","    # Highlight specific combinations in red\n","    for t, sp, comb in zip(patient_ts, patient_significant_p_values, patient_combinations):\n","        if comb in highlight_set:\n","            ax.plot(t, sp, 'ro')\n","\n","    ax.set_xlabel('Batch')\n","    ax.set_ylabel('Number of non stationnary species')\n","    ax.set_title(f'Patient {patient}')\n","    ax.grid(True, linestyle='--', alpha=0.7)\n","\n","# Remove empty subplots\n","for i in range(n, rows*cols):\n","    row = i // cols\n","    col = i % cols\n","    fig.delaxes(axes[row, col] if rows > 1 else axes[col])\n","\n","plt.tight_layout()\n","plt.subplots_adjust(top=0.95)\n","plt.show()\n"],"metadata":{"id":"nG9cu8qHoa82","executionInfo":{"status":"aborted","timestamp":1723144946836,"user_tz":240,"elapsed":4161,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4161,"status":"aborted","timestamp":1723144946836,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"M_JMXhjCzz23"},"outputs":[],"source":["analyze_time_series(gene_count_df, 0.1, \"bh\")"]},{"cell_type":"code","source":["# @title copy paste output here\n","\n","output = \"\"\"\n","Length: 97, dtype: int64\n","Patient ID: TR_2101_TS2\n","Number of variables with significant adjusted p-values: 162\n","Patient ID: TR_2101_TS3\n","Number of variables with significant adjusted p-values: 119\n","Patient ID: TR_2101_TS4\n","Number of variables with significant adjusted p-values: 120\n","Patient ID: TR_2101_TS5\n","Number of variables with significant adjusted p-values: 125\n","Patient ID: TR_2101_TS6\n","Number of variables with significant adjusted p-values: 147\n","Patient ID: TR_2101_TS7\n","Number of variables with significant adjusted p-values: 66\n","Patient ID: TR_2101_TS8\n","Number of variables with significant adjusted p-values: 98\n","Patient ID: TR_2102_TS1\n","Number of variables with significant adjusted p-values: 173\n","Patient ID: TR_2102_TS2\n","Number of variables with significant adjusted p-values: 124\n","Patient ID: TR_2102_TS3\n","Number of variables with significant adjusted p-values: 102\n","Patient ID: TR_2102_TS4\n","Number of variables with significant adjusted p-values: 127\n","Patient ID: TR_2102_TS5\n","Number of variables with significant adjusted p-values: 133\n","Patient ID: TR_2102_TS6\n","Number of variables with significant adjusted p-values: 100\n","Patient ID: TR_2102_TS7\n","Number of variables with significant adjusted p-values: 46\n","Patient ID: TR_2103_TS2\n","Number of variables with significant adjusted p-values: 107\n","Patient ID: TR_2103_TS3\n","Number of variables with significant adjusted p-values: 107\n","Patient ID: TR_2103_TS4\n","Number of variables with significant adjusted p-values: 91\n","Patient ID: TR_2103_TS5\n","Number of variables with significant adjusted p-values: 165\n","Patient ID: TR_2103_TS6\n","Number of variables with significant adjusted p-values: 126\n","Patient ID: TR_2103_TS7\n","Number of variables with significant adjusted p-values: 135\n","Patient ID: TR_2103_TS8\n","Number of variables with significant adjusted p-values: 165\n","Patient ID: TR_2104_TS1\n","Number of variables with significant adjusted p-values: 73\n","Patient ID: TR_2104_TS2\n","Number of variables with significant adjusted p-values: 150\n","Patient ID: TR_2104_TS3\n","Number of variables with significant adjusted p-values: 78\n","Patient ID: TR_2104_TS4\n","Number of variables with significant adjusted p-values: 101\n","Patient ID: TR_2104_TS5\n","Number of variables with significant adjusted p-values: 119\n","Patient ID: TR_2104_TS6\n","Number of variables with significant adjusted p-values: 147\n","Patient ID: TR_2104_TS7\n","Number of variables with significant adjusted p-values: 122\n","Patient ID: TR_2104_TS8\n","Number of variables with significant adjusted p-values: 142\n","Patient ID: TR_2105_TS1\n","Number of variables with significant adjusted p-values: 84\n","Patient ID: TR_2105_TS2\n","Number of variables with significant adjusted p-values: 111\n","Patient ID: TR_2105_TS3\n","Number of variables with significant adjusted p-values: 126\n","Patient ID: TR_2105_TS4\n","Number of variables with significant adjusted p-values: 89\n","Patient ID: TR_2105_TS5\n","Number of variables with significant adjusted p-values: 99\n","Patient ID: TR_2105_TS6\n","Number of variables with significant adjusted p-values: 115\n","Patient ID: TR_2105_TS7\n","Number of variables with significant adjusted p-values: 150\n","Patient ID: TR_2105_TS8\n","Number of variables with significant adjusted p-values: 103\n","Patient ID: TR_2106_TS1\n","Number of variables with significant adjusted p-values: 165\n","Patient ID: TR_2106_TS2\n","Number of variables with significant adjusted p-values: 165\n","Patient ID: TR_2106_TS3\n","Number of variables with significant adjusted p-values: 152\n","Patient ID: TR_2106_TS4\n","Number of variables with significant adjusted p-values: 173\n","Patient ID: TR_2106_TS5\n","Number of variables with significant adjusted p-values: 115\n","Patient ID: TR_2106_TS6\n","Number of variables with significant adjusted p-values: 145\n","Patient ID: TR_2106_TS7\n","Number of variables with significant adjusted p-values: 126\n","Patient ID: TR_2106_TS8\n","Number of variables with significant adjusted p-values: 161\n","Patient ID: TR_2107_TS2\n","Number of variables with significant adjusted p-values: 182\n","Patient ID: TR_2107_TS3\n","Number of variables with significant adjusted p-values: 180\n","Patient ID: TR_2107_TS4\n","Number of variables with significant adjusted p-values: 103\n","Patient ID: TR_2107_TS5\n","Number of variables with significant adjusted p-values: 108\n","Patient ID: TR_2107_TS6\n","Number of variables with significant adjusted p-values: 130\n","Patient ID: TR_2107_TS7\n","Number of variables with significant adjusted p-values: 145\n","Patient ID: TR_2107_TS8\n","Number of variables with significant adjusted p-values: 173\n","Patient ID: TR_2108_TS1\n","Number of variables with significant adjusted p-values: 159\n","Patient ID: TR_2108_TS2\n","Number of variables with significant adjusted p-values: 196\n","Patient ID: TR_2108_TS3\n","Number of variables with significant adjusted p-values: 161\n","Patient ID: TR_2108_TS4\n","Number of variables with significant adjusted p-values: 128\n","Patient ID: TR_2108_TS5\n","Number of variables with significant adjusted p-values: 177\n","Patient ID: TR_2108_TS6\n","Number of variables with significant adjusted p-values: 98\n","Patient ID: TR_2201_TS1\n","Number of variables with significant adjusted p-values: 94\n","Patient ID: TR_2201_TS2\n","Number of variables with significant adjusted p-values: 136\n","Patient ID: TR_2201_TS3\n","Number of variables with significant adjusted p-values: 142\n","Patient ID: TR_2201_TS4\n","Number of variables with significant adjusted p-values: 140\n","Patient ID: TR_2201_TS5\n","Number of variables with significant adjusted p-values: 145\n","Patient ID: TR_2201_TS6\n","Number of variables with significant adjusted p-values: 76\n","Patient ID: TR_2201_TS7\n","Number of variables with significant adjusted p-values: 111\n","Patient ID: TR_2201_TS8\n","Number of variables with significant adjusted p-values: 61\n","Patient ID: TR_2202_TS1\n","Number of variables with significant adjusted p-values: 108\n","Patient ID: TR_2202_TS2\n","Number of variables with significant adjusted p-values: 104\n","Patient ID: TR_2202_TS3\n","Number of variables with significant adjusted p-values: 94\n","Patient ID: TR_2202_TS4\n","Number of variables with significant adjusted p-values: 138\n","Patient ID: TR_2202_TS5\n","Number of variables with significant adjusted p-values: 101\n","Patient ID: TR_2202_TS6\n","Number of variables with significant adjusted p-values: 145\n","Patient ID: TR_2202_TS7\n","Number of variables with significant adjusted p-values: 93\n","Patient ID: TR_2202_TS8\n","Number of variables with significant adjusted p-values: 94\n","Patient ID: TR_2203_TS1\n","Number of variables with significant adjusted p-values: 56\n","Patient ID: TR_2203_TS2\n","Number of variables with significant adjusted p-values: 136\n","Patient ID: TR_2203_TS3\n","Number of variables with significant adjusted p-values: 100\n","Patient ID: TR_2203_TS4\n","Number of variables with significant adjusted p-values: 94\n","Patient ID: TR_2203_TS5\n","Number of variables with significant adjusted p-values: 89\n","Patient ID: TR_2203_TS6\n","Number of variables with significant adjusted p-values: 102\n","Patient ID: TR_2203_TS7\n","Number of variables with significant adjusted p-values: 82\n","Patient ID: TR_2203_TS8\n","Number of variables with significant adjusted p-values: 72\n","Patient ID: TR_2204_TS1\n","Number of variables with significant adjusted p-values: 196\n","Patient ID: TR_2204_TS2\n","Number of variables with significant adjusted p-values: 104\n","Patient ID: TR_2204_TS3\n","Number of variables with significant adjusted p-values: 116\n","Patient ID: TR_2204_TS4\n","Number of variables with significant adjusted p-values: 124\n","Patient ID: TR_2204_TS5\n","Number of variables with significant adjusted p-values: 167\n","Patient ID: TR_2204_TS6\n","Number of variables with significant adjusted p-values: 143\n","Patient ID: TR_2204_TS7\n","Number of variables with significant adjusted p-values: 149\n","Patient ID: TR_2205_TS1\n","Number of variables with significant adjusted p-values: 84\n","Patient ID: TR_2205_TS2\n","Number of variables with significant adjusted p-values: 140\n","Patient ID: TR_2205_TS3\n","Number of variables with significant adjusted p-values: 165\n","Patient ID: TR_2205_TS4\n","Number of variables with significant adjusted p-values: 149\n","Patient ID: TR_2205_TS5\n","Number of variables with significant adjusted p-values: 138\n","Patient ID: TR_2205_TS6\n","Number of variables with significant adjusted p-values: 159\n","Patient ID: TR_2205_TS7\n","Number of variables with significant adjusted p-values: 129\n","Patient ID: TR_2205_TS8\n","Number of variables with significant adjusted p-values: 213\n","\"\"\"\n","\n"],"metadata":{"cellView":"form","id":"6XDR8gOJ85-v","executionInfo":{"status":"aborted","timestamp":1723144946836,"user_tz":240,"elapsed":4161,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["patients  = []\n","ts  = []\n","significant_p_values = []\n","\n","# Use regular expressions to extract the information\n","pattern = re.compile(r\"Patient ID: TR_(\\d+)_TS(\\d+)\\s+Number of variables with significant adjusted p-values: (\\d+)\")\n","\n","matches = pattern.findall(output)\n","\n","for match in matches:\n","    patients_, ts_, significant_p_values_ = match\n","    patients.append(int(patients_))\n","    ts.append(int(ts_))\n","    significant_p_values.append(int(significant_p_values_))"],"metadata":{"id":"TnNnKtT69KSQ","executionInfo":{"status":"aborted","timestamp":1723144946837,"user_tz":240,"elapsed":4162,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["highlight_set = set(metaphlan_df[metaphlan_df[\"RBF\"]==True]['ts_patient_id'].unique())\n","\n","# Generate combinations of Patient and TS\n","combinations = ['TR_' + str(p) + '_TS' + str(t) for p, t in zip(patients, ts)]\n","\n","# Get unique patients\n","unique_patients = sorted(set(patients))\n","\n","# Calculate the number of rows and columns for subplots\n","n = len(unique_patients)\n","cols = 3\n","rows = math.ceil(n / cols)\n","\n","# Create the plot\n","fig, axes = plt.subplots(rows, cols, figsize=(15, 5*rows))\n","\n","for i, patient in enumerate(unique_patients):\n","    row = i // cols\n","    col = i % cols\n","    ax = axes[row, col] if rows > 1 else axes[col]\n","\n","    # Filter data for the current patient\n","    patient_indices = [j for j, p in enumerate(patients) if p == patient]\n","    patient_ts = [ts[j] for j in patient_indices]\n","    patient_significant_p_values = [significant_p_values[j] for j in patient_indices]\n","    patient_combinations = [combinations[j] for j in patient_indices]\n","\n","    # Plot the data\n","    ax.plot(patient_ts, patient_significant_p_values, label=f'Patient {patient}')\n","\n","    # Highlight specific combinations in red\n","    for t, sp, comb in zip(patient_ts, patient_significant_p_values, patient_combinations):\n","        if comb in highlight_set:\n","            ax.plot(t, sp, 'ro')\n","\n","    ax.set_xlabel('Batch')\n","    ax.set_ylabel('Number of non stationnary genes')\n","    ax.set_title(f'Patient {patient}')\n","    ax.grid(True, linestyle='--', alpha=0.7)\n","\n","# Remove empty subplots\n","for i in range(n, rows*cols):\n","    row = i // cols\n","    col = i % cols\n","    fig.delaxes(axes[row, col] if rows > 1 else axes[col])\n","\n","plt.tight_layout()\n","plt.subplots_adjust(top=0.95)\n","plt.show()\n"],"metadata":{"id":"OOCD_56XsrF9","executionInfo":{"status":"aborted","timestamp":1723144946837,"user_tz":240,"elapsed":4162,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4445,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"attoxCMQOEHQ"},"outputs":[],"source":["# @title shifting the week numbers so that the three patients have the beginning of their flare at the same time, which makes it easier to compare visually\n","metaphlan_df = align_flare_start_week(metaphlan_df)\n","gene_count_df = align_flare_start_week(gene_count_df)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4445,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"BZbkjO7BksVn"},"outputs":[],"source":["species_signif_in_adf = [\"s__Bacteroides_ovatus\", \"s__Barnesiella_intestinihominis\", \"s__Alistipes_onderdonkii\", \"s__Clostridiaceae_bacterium\", \"s__Anaerostipes_hadrus\", \"s__Blautia_wexlerae\", \"s__Dorea_formicigenerans\", \"s__Dorea_longicatena\", \"s__Eubacterium_rectale\", \"s__Lachnospiraceae_bacterium\", \"s__Roseburia_inulinivorans\", \"s__Flavonifractor_plautii\", \"s__Ruminococcus_bicirculans\", \"s__Faecalibacillus_intestinalis\", \"s__Bacteroides_uniformis\", \"s__Parabacteroides_distasonis\", \"s__Agathobaculum_butyriciproducens\", \"s__Faecalibacterium_prausnitzii\"]\n"]},{"cell_type":"code","source":["len(species_signif_in_adf)"],"metadata":{"id":"i3tqLIO8U3ix","executionInfo":{"status":"aborted","timestamp":1723144947120,"user_tz":240,"elapsed":4445,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["species_signif_in_adf[9:17]"],"metadata":{"id":"ec8wNs5WVFrB","executionInfo":{"status":"aborted","timestamp":1723144947120,"user_tz":240,"elapsed":4444,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4444,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"5FD7x_3mkzBK"},"outputs":[],"source":["plot_species_abundance_over_time_2(metaphlan_df, cols_to_plot=species_signif_in_adf[], title='Abundance', type_='Species')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9Y0aRMnYlDYM","executionInfo":{"status":"aborted","timestamp":1723144947120,"user_tz":240,"elapsed":4444,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"outputs":[],"source":["genes_signif_in_adf = [\"ACVR1B_ENSG00000135503_ENST00000257963\", \"APAF1_ENSG00000120868_ENST00000551964\", \"BMP2_ENSG00000125845_ENST00000378827\", \"CASP3_ENSG00000164305_ENST00000308394\", \"CCL20_ENSG00000115009_ENST00000358813\", \"CLDN8_ENSG00000156284_ENST00000399899\", \"CLEC7A_ENSG00000172243_ENST00000304084\", \"CXCL1_ENSG00000163739_ENST00000395761\", \"FCGBP_ENSG00000275395_ENST00000616721\", \"FFAR2_ENSG00000126262_ENST00000599180\", \"HMGB1_ENSG00000189403_ENST00000341423\", \"ITGB1_ENSG00000150093_ENST00000437302\", \"MMP12_ENSG00000262406_ENST00000571244\", \"NR1D2_ENSG00000174738_ENST00000312521\", \"REL_ENSG00000162924_ENST00000394479\", \"RIPK2_ENSG00000104312_ENST00000220751\", \"S100A9_ENSG00000163220_ENST00000368738\", \"SELENBP1_ENSG00000143416_ENST00000368868\", \"TNFRSF25_ENSG00000215788_ENST00000356876\", \"XBP1_ENSG00000100219_ENST00000344347\", \"ATG5_ENSG00000057663_ENST00000369076\", \"CD68_ENSG00000129226_ENST00000250092\", \"CXCR5_ENSG00000160683_ENST00000292174\", \"DEFB1_ENSG00000164825_ENST00000297439\", \"IL4R_ENSG00000077238_ENST00000395762\", \"IRAG2_ENSG00000118308_ENST00000636465\", \"NFKB1_ENSG00000109320_ENST00000226574\", \"PIK3CA_ENSG00000121879_ENST00000263967\", \"TAGAP_ENSG00000164691_ENST00000367066\", \"BCAR1_ENSG00000050820_ENST00000162330\", \"CD44_ENSG00000026508_ENST00000428726\", \"IL1R2_ENSG00000115590_ENST00000332549\", \"NCF2_ENSG00000116701_ENST00000367535\", \"PTEN_ENSG00000171862_ENST00000371953\", \"SATB2_ENSG00000119042_ENST00000417098\", \"TNFRSF14_ENSG00000157873_ENST00000355716\", \"NOD2_ENSG00000167207_ENST00000647318\"]\n"]},{"cell_type":"code","source":["len(genes_signif_in_adf)"],"metadata":{"id":"7IdajDhzUwoq","executionInfo":{"status":"aborted","timestamp":1723144947120,"user_tz":240,"elapsed":4444,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["18+37"],"metadata":{"id":"YIw06On9Va4z","executionInfo":{"status":"aborted","timestamp":1723144947120,"user_tz":240,"elapsed":4444,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# @ title species to investigate\n","plot_species_abundance_over_time_2(gene_count_df,   genes_signif_in_adf, 'Expression', 'Genes')"],"metadata":{"id":"6heNQgzImB8h","executionInfo":{"status":"aborted","timestamp":1723144947120,"user_tz":240,"elapsed":4444,"user":{"displayName":"Clémence","userId":"00304126927071239250"}}},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4444,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"D2h6DAc4DuR6"},"outputs":[],"source":["# @ title species to investigate\n","plot_species_abundance_over_time_2(gene_count_df,   ['IL17A_ENSG00000112115_ENST00000648244', 'MKI67_ENSG00000148773_ENST00000368654', 'PRLR_ENSG00000113494_ENST00000618457'], 'Expression', 'Genes')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4444,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"1dF03sIeng-F"},"outputs":[],"source":["plot_species_abundance_over_time_2(metaphlan_df, cols_to_plot=['s__Flavonifractor_plautii'], title='Abundance', type_='Species')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4444,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"4qlNxU9_z0g3"},"outputs":[],"source":["plot_species_abundance_over_time_2(metaphlan_df, cols_to_plot=['s__Barnesiella_sp_An22','s__Gordonibacter_pamelaeae','s__Firmicutes_bacterium_AF16_15'], title='Abundance', type_='Species')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4444,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"dNP049Wp_cYQ"},"outputs":[],"source":["plot_species_abundance_over_time(metaphlan_df, ['s__Flavonifractor_plautii', 's__Blautia_wexlerae'], 'Microbial Abundance', 'Taxa')"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","executionInfo":{"elapsed":4444,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"YuVq3_PA51Sj"},"outputs":[],"source":["# @title Aggregate species\n","\n","# Extract the genus part from the species column names\n","metaphlan_agg_col = {col: col.split('_')[2] for col in metaphlan_df.columns if 's__' in col}\n","\n","# Map species to genus\n","metaphlan_agg = metaphlan_df.rename(columns=metaphlan_agg_col)\n","\n","metaphlan_agg = metaphlan_agg.T.groupby(level=0).sum().T\n","\n","exclude_columns  = [ 'PatientID_Weeknr',\n"," 'patient_id',\n"," 'week',\n"," 'Flare_status',\n"," 'Flare_start',\n"," 'ts',\n"," 'RBF',\n"," 'ts_patient_id']\n","\n","new_order =  exclude_columns + [col for col in metaphlan_agg.columns if col not in exclude_columns]\n","metaphlan_agg = metaphlan_agg[new_order]\n","\n","# Your DataFrame now has the microbial data aggregated by genus\n","print(metaphlan_agg.head())\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"9x8nWPrZeWbU"},"outputs":[],"source":["analyze_representation_and_variation(metaphlan_agg, 8, 'Taxa')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"GArebV1Y5P9y"},"outputs":[],"source":["plot_species_abundance_over_time(metaphlan_agg, ['Bacteroides', 'Christensenellaceae', 'Gemella', 'Sutterella'], 'Microbial Abundance', 'Taxa')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"3YTP-uh6HvGY"},"outputs":[],"source":["plot_species_abundance_over_time_2(gene_count_df,   ['XCR1_ENSG00000173578_ENST00000683768'], 'Expression', 'Genes')"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"bjE0hb2iH8mZ"},"outputs":[],"source":["metaphlan_df.columns"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"dUQMnkgiIVJR"},"outputs":[],"source":["gene_count_df.columns"]},{"cell_type":"markdown","metadata":{"id":"KRTiWiXkasAf"},"source":["\n","\n","# Smoothed Profiles Analysis\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"5c5eSPkmdzQw"},"outputs":[],"source":["species_columns = [col for col in metaphlan_df.columns if col not in exclude_columns]\n","gene_columns = [col for col in gene_count_df.columns if col not in exclude_columns]"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"JkUeuSWEd4GA"},"outputs":[],"source":["smoothed_metaphlan_df = smooth_patient_data(metaphlan_df, species_columns)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4443,"status":"aborted","timestamp":1723144947120,"user":{"displayName":"Clémence","userId":"00304126927071239250"},"user_tz":240},"id":"1QiY4KRKeAUb"},"outputs":[],"source":["plot_species_abundance_over_time_2(smoothed_metaphlan_df , cols_to_plot=['s__Phocaeicola_dorei','s__Bacteroides_uniformis','s__Faecalibacterium_prausnitzii' , 's__Bifidobacterium_pullorum',\n","  's__Bacteroides_ovatus', 's__Clostridia_bacterium', 's__Eubacterium_rectale', 's__Blautia_sp_MSK_21_1',\n","  's__GGB4569_SGB6310', 's__Blautia_wexlerae'], title='Abundance', type_='Species')"]}],"metadata":{"colab":{"provenance":[{"file_id":"1vKJuzEWFSFrv7Aj5iBjy09DdM7_i_SbK","timestamp":1716406919952},{"file_id":"1izB2erfb3QLGUppLrFFLhlczRq8qFxrN","timestamp":1715284665173},{"file_id":"1UAW5rTXSCVmFM96mOpjEtZiQNsMGjYy9","timestamp":1715205490615},{"file_id":"1INX8yiEhMiTnnEh0mbt4wsHhHGX49FXn","timestamp":1713368007132}]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}